memcached/memcached_1.6.19.patch

```diff
@@ -0,0 +1,1308 @@
+ 
+@@ -2726,7 +2729,7 @@ static void process_lru_crawler_command(conn *c, token_t *tokens, const size_t n
+         out_string(c, "ERROR");
+     }
+ }
+-#ifdef TLS
++#if defined(TLS) || defined(WOLFSSL_MEMCACHED)
+ static void process_refresh_certs_command(conn *c, token_t *tokens, const size_t ntokens) {
+     set_noreply_maybe(c, tokens, ntokens);
+     char *errmsg = NULL;
+@@ -2948,7 +2951,7 @@ void process_command_ascii(conn *c, char *command) {
+         WANT_TOKENS_MIN(ntokens, 3);
+         process_extstore_command(c, tokens, ntokens);
+ #endif
+-#ifdef TLS
++#if defined(TLS) || defined(WOLFSSL_MEMCACHED)
+     } else if (strcmp(tokens[COMMAND_TOKEN].value, "refresh_certs") == 0) {
+         process_refresh_certs_command(c, tokens, ntokens);
+ #endif
+diff --git a/testapp.c b/testapp.c
+index 387a847..393adcf 100644
+--- a/testapp.c
++++ b/testapp.c
+@@ -29,6 +29,13 @@
+ #ifdef TLS
+ #include <openssl/ssl.h>
+ #endif
++#ifdef WOLFSSL_MEMCACHED
++#ifndef WOLFSSL_USER_SETTINGS
++#include <wolfssl/options.h>
++#endif
++#include <wolfssl/wolfcrypt/settings.h>
++#include <wolfssl/ssl.h>
++#endif
+ 
+ #define TMP_TEMPLATE "/tmp/test_file.XXXXXXX"
+ 
+@@ -39,6 +46,10 @@ struct conn {
+ #ifdef TLS
+     SSL_CTX   *ssl_ctx;
+     SSL    *ssl;
++#endif
++#ifdef WOLFSSL_MEMCACHED
++    WOLFSSL_CTX *ssl_ctx;
++    WOLFSSL     *ssl;
+ #endif
+     ssize_t (*read)(struct conn  *c, void *buf, size_t count);
+     ssize_t (*write)(struct conn *c, const void *buf, size_t count);
+@@ -48,7 +59,7 @@ hash_func hash;
+ 
+ static ssize_t tcp_read(struct conn *c, void *buf, size_t count);
+ static ssize_t tcp_write(struct conn *c, const void *buf, size_t count);
+-#ifdef TLS
++#if defined(TLS) || defined(WOLFSSL_MEMCACHED)
+ static ssize_t ssl_read(struct conn *c, void *buf, size_t count);
+ static ssize_t ssl_write(struct conn *c, const void *buf, size_t count);
+ #endif
+@@ -73,6 +84,31 @@ ssize_t ssl_write(struct conn *c, const void *buf, size_t count) {
+     return SSL_write(c->ssl, buf, count);
+ }
+ #endif
++#ifdef WOLFSSL_MEMCACHED
++ssize_t ssl_read(struct conn *c, void *buf, size_t count) {
++    int ret = -1;
++
++    assert(c != NULL);
++    ret = wolfSSL_read(c->ssl, buf, count);
++    if (ret < 0) {
++        ret = wolfSSL_get_error(c->ssl, ret);
++    }
++
++    return ret;
++}
++
++ssize_t ssl_write(struct conn *c, const void *buf, size_t count) {
++    int ret = -1;
++
++    assert(c != NULL);
++    ret = wolfSSL_write(c->ssl, buf, count);
++    if (ret < 0) {
++        ret = wolfSSL_get_error(c->ssl, ret);
++    }
++
++    return ret;
++}
++#endif
+ 
+ static pid_t server_pid;
+ static in_port_t port;
+@@ -89,6 +125,14 @@ static void close_conn(void) {
+     }
+     if (con->ssl_ctx)
+         SSL_CTX_free(con->ssl_ctx);
++#endif
++#ifdef WOLFSSL_MEMCACHED
++    if (con->ssl) {
++        wolfSSL_shutdown(con->ssl);
++        wolfSSL_free(con->ssl);
++    }
++    if (con->ssl_ctx)
++        wolfSSL_CTX_free(con->ssl_ctx);
```

Comment:
Why is this not using the OpenSSL macros?

Reply:
Judging by the `configure.ac` it looks like the compiler should be able to pick up the openssl macros.

Reply:
Since you include `-I$WOLFSSL_INCLUDEDIR/wolfssl` you don't need to change the includes either since it will pick up the compatibility headers instead of the real ossl ones.